### PR TITLE
fix:[CORE-1833] Insert audit logs when dangling issues are closed

### PR DESCRIPTION
**Pull request by ershad-paladin**
## Description

- issue: audit log entry is missed when dangling issues are closed by policy executor
- fix: made  AuditUtils.postAuditTrail generic and used the same common method to insert audit log when dangling issues are closed

### Problem
audit log entry is missed when dangling issues are closed by policy executor

### Solution
made  AuditUtils.postAuditTrail generic and used the same common method to insert audit log when dangling issues are closed

## Type of change

**Please delete options that are not relevant.**

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Chore (no code changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also
list any relevant details for your test configuration

- [X] Replicate the issue first by disabling the policy 'Check for Publicly Accessible Cloud KMS Keys' and verify there is no audit log with status = closed
- [X] enable the same policy again, apply fix and disable the policy
- [X] verify there is an audit log with status = closed

image
image

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] My commit message/PR follows the contribution guidelines of this project
- [X] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

- **New Features**
  - Added a new constant `POLICY_DISABLED_MSG` to communicate disabled policy status.

- **Improvements**
  - Enhanced auditing by updating methods in `AuditUtils` to accept map structures instead of specific objects.

- **Bug Fixes**
  - Improved closing of dangling issues in `PolicyExecutor` with a new condition based on policy status.

- **Refactor**
  - Updated method signatures in `AuditUtils` and `CommonUtils` for accepting map structures.

### CODE REVIEW

**Review by AnandPaladin — COMMENTED** [review]
@ershad-paladin this will work. Nice approach. But based upon my analysis , can we leverage **existing method call  AuditUtils.postAuditTrail(allIssues, originalIssueStatuses)** at PolicyExecutor at Line Number 603 by adding "closedDanglingIssuesList" in to allIssues list before calling existing above method . Ofcourse that will require change of return of closeDanglingIssues from Integer to Map and much more changes. Just wanted to know your thoughts. Please see the below screen shot what I thought .Left side existing code right side my proposed changes. Of course my approach seems to be having more code than you,
image
CC @santhosh-challa

**Review by plyubich — CHANGES_REQUESTED** [review]
Please see comments

File: jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/executor/PolicyExecutor.java [hunk 1]
```diff
@@ -120,12 +120,13 @@ public static void main(String[] args) {
...
                     if (policyExecutor.resources.isEmpty()
                             || PacmanSdkConstants.POLICY_STATUS_DISABLED.equalsIgnoreCase(status)) {
+                        String reasonToClose = PacmanSdkConstants.POLICY_STATUS_DISABLED.equalsIgnoreCase(status) ? "Policy has been disabled" : PacmanSdkConstants.REASON_TO_CLOSE_VALUE;
                         AnnotationPublisher annotationPublisher = new AnnotationPublisher();
                         annotationPublisher.populateExistingIssuesForType(policyParam);
                         annotationPublisher.closeDanglingIssues(
                                 policyParam.get(PacmanSdkConstants.DATA_SOURCE_KEY) + "_"
                                         + policyParam.get(PacmanSdkConstants.TARGET_TYPE),
-                                policyParam.get(PacmanSdkConstants.TARGET_TYPE));
+                                reasonToClose);
```
[the remark this answers is not in the document]
Comment (ershad-paladin) [reply jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/executor/PolicyExecutor.java:130]: done

File: jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/executor/PolicyExecutor.java [hunk 2]
```diff
@@ -592,7 +593,7 @@ private Map<String, Object> processPolicyEvaluations(List<Map<String, String>> r
         NotificationUtils.triggerNotificationsForViolations(bulkUploadAnnotations, annotationPublisher.getExistingIssuesMapWithAnnotationIdAsKey(), true);
         NotificationUtils.triggerNotificationsForViolations(annotationPublisher.getClouserBucket(), annotationPublisher.getExistingIssuesMapWithAnnotationIdAsKey(), false);
 
-        Integer danglisngIssues = annotationPublisher.closeDanglingIssues(annotation);
+        Integer danglisngIssues = annotationPublisher.closeDanglingIssues(annotation, PacmanSdkConstants.REASON_TO_CLOSE_VALUE);
```
[the remark this answers is not in the document]
Comment (ershad-paladin) [reply jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/executor/PolicyExecutor.java:589]: > Please remove above commented code

done

> Shouldn't you propagate your reasonToClose from the beginning here as well?

This is the beginning where we are closing the issues for which the asset is no more, after processing the open issues

File: jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/publisher/impl/AnnotationPublisher.java [hunk 3]
```diff
@@ -394,11 +393,9 @@ public List<Annotation> processClosureEx() throws Exception {
      * @return the int
      * @throws Exception the exception
      */
-    public int closeDanglingIssues(Annotation sampleAnnotation) throws Exception {
+    public int closeDanglingIssues(Annotation sampleAnnotation, String reason) throws Exception {
```
[the remark this answers is not in the document]
Comment (ershad-paladin) [reply jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/publisher/impl/AnnotationPublisher.java:396]: one of the annotations from a list is passed into this method just to provide some generic annotation metadata like index name

File: jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/util/AuditUtils.java [hunk 4]
```diff
@@ -131,14 +133,17 @@ private static String createAuditTrail(String ds, String type, String status, St
...
         auditTrail.put(TARGET_TYPE, type);
         auditTrail.put("_docid",id);
         auditTrail.put(PacmanSdkConstants.ANNOTATION_PK, id);
         auditTrail.put(PacmanSdkConstants.STATUS_KEY, status);
         auditTrail.put(PacmanSdkConstants.AUDIT_DATE, date);
         auditTrail.put(PacmanSdkConstants._AUDIT_DATE, date.substring(0, date.indexOf('T')));
         auditTrail.put(CREATED_BY, SYSTEM);
+        if (STATUS_CLOSE.equalsIgnoreCase(status) && StringUtils.isNotBlank(annotation.get(PacmanSdkConstants.REASON_TO_CLOSE_KEY))) {
+            auditTrail.put(PacmanSdkConstants.REASON_TO_CLOSE_KEY, annotation.get(PacmanSdkConstants.REASON_TO_CLOSE_KEY));
+        }
```
[the remark this answers is not in the document]
Comment (ershad-paladin) [reply jobs/pacman-rule-engine-2.0/src/main/java/com/tmobile/pacman/util/AuditUtils.java:146]: we are not missing the audit log. we are setting the reason-to-close field in the audit log if the violation has one

**Comment by ershad-paladin** [comment]
tested screenshot:
image

**Review by plyubich — APPROVED** [review]

**Review by santhosh-challa — APPROVED** [review]

**Pull request merged**